File: packages/pp_core/lib/src/einvoice_csv_parser.dart
import 'package:csv/csv.dart';

/// One parsed invoice (header + line items) from a Ministry of Finance carrier
/// "消費明細" (detail) CSV export.
class ParsedInvoice {
  final String invoiceNumber;
  final DateTime date;
  final String? merchantName; // 賣方名稱
  final String? carrierName; // 載具自訂名稱
  final String? sellerAddress; // 賣方地址
  final List<ParsedItem> items;

  ParsedInvoice({
    required this.invoiceNumber,
    required this.date,
    this.merchantName,
    this.carrierName,
    this.sellerAddress,
    required this.items,
  });

  /// Invoice total in **TWD dollars** = sum of line amounts.
  ///
  /// The CSV's 發票金額 column is unreliable — for multi-line invoices it
  /// mirrors the per-line amount rather than the invoice total — so we always
  /// sum the line items (discount lines are negative and net correctly).
  int get totalDollars => items.fold(0, (sum, i) => sum + i.amount);
}

/// A single line item, amounts in **TWD dollars** (the CSV is not in cents).
class ParsedItem {
  final String name; // 消費明細_品名
  final num quantity; // 消費明細_數量
  final int unitPrice; // 消費明細_單價
  final int amount; // 消費明細_金額

  const ParsedItem({
    required this.name,
    required this.quantity,
    required this.unitPrice,
    required this.amount,
  });
}

// 2 letters + 8 digits (digits may be masked with '*' for voided/donated ones).
final _invoiceNoRe = RegExp(r'^[A-Z]{2}[0-9*]{8}$');
final _dateRe = RegExp(r'^\d{8}$');

/// Parses a MOF carrier detail CSV into invoices grouped by invoice number.
///
/// Robust to the file's quirks: a UTF-8 BOM, CRLF line endings, the header row,
/// and the 2 trailing footnote lines (all skipped because only rows with a
/// valid invoice number + YYYYMMDD date are accepted). Columns are located by
/// header name, falling back to fixed positions if the header is missing.
List<ParsedInvoice> parseEinvoiceCsv(String csv) {
  var text = csv;
  if (text.isNotEmpty && text.codeUnitAt(0) == 0xFEFF) {
    text = text.substring(1); // strip UTF-8 BOM
  }
  text = text.replaceAll('\r\n', '\n').replaceAll('\r', '\n');

  final rows = const CsvToListConverter(eol: '\n', shouldParseNumbers: false)
      .convert(text);
  if (rows.isEmpty) return const [];

  final col = _columnIndex(rows);

  // Preserve first-seen order while grouping by invoice number.
  final order = <String>[];
  final groups = <String, _InvoiceAccumulator>{};

  for (final row in rows) {
    if (row.length <= col.item) continue;
    final no = _cell(row, col.number);
    final dateStr = _cell(row, col.date);
    if (!_invoiceNoRe.hasMatch(no) || !_dateRe.hasMatch(dateStr)) {
      continue; // header, footnotes, or any misaligned row
    }

    final acc = groups.putIfAbsent(no, () {
      order.add(no);
      return _InvoiceAccumulator(
        invoiceNumber: no,
        date: _parseDate(dateStr),
        merchantName: _nullIfEmpty(_cell(row, col.merchant)),
        carrierName: _nullIfEmpty(_cell(row, col.carrier)),
        sellerAddress: _nullIfEmpty(_cell(row, col.address)),
      );
    });

    final name = _cell(row, col.item);
    if (name.isEmpty) continue;
    acc.items.add(ParsedItem(
      name: name,
      quantity: num.tryParse(_cell(row, col.qty)) ?? 1,
      unitPrice: _toInt(_cell(row, col.unitPrice)),
      amount: _toInt(_cell(row, col.amount)),
    ));
  }

  return [
    for (final no in order)
      if (groups[no]!.items.isNotEmpty) groups[no]!.build(),
  ];
}

class _Columns {
  final int carrier, date, number, merchant, address, qty, unitPrice, amount,
      item;
  const _Columns(this.carrier, this.date, this.number, this.merchant,
      this.address, this.qty, this.unitPrice, this.amount, this.item);
}

_Columns _columnIndex(List<List<dynamic>> rows) {
  final byName = <String, int>{};
  for (final row in rows) {
    if (row.any((c) => c.toString().trim() == '發票號碼') &&
        row.any((c) => c.toString().trim() == '消費明細_品名')) {
      for (var i = 0; i < row.length; i++) {
        byName[row[i].toString().trim()] = i;
      }
      break;
    }
  }
  int idx(String name, int fallback) => byName[name] ?? fallback;
  return _Columns(
    idx('載具自訂名稱', 0),
    idx('發票日期', 1),
    idx('發票號碼', 2),
    idx('賣方名稱', 7),
    idx('賣方地址', 8),
    idx('消費明細_數量', 10),
    idx('消費明細_單價', 11),
    idx('消費明細_金額', 12),
    idx('消費明細_品名', 13),
  );
}

class _InvoiceAccumulator {
  final String invoiceNumber;
  final DateTime date;
  final String? merchantName;
  final String? carrierName;
  final String? sellerAddress;
  final List<ParsedItem> items = [];

  _InvoiceAccumulator({
    required this.invoiceNumber,
    required this.date,
    this.merchantName,
    this.carrierName,
    this.sellerAddress,
  });

  ParsedInvoice build() => ParsedInvoice(
        invoiceNumber: invoiceNumber,
        date: date,
        merchantName: merchantName,
        carrierName: carrierName,
        sellerAddress: sellerAddress,
        items: items,
      );
}

String _cell(List<dynamic> row, int i) =>
    i < row.length ? row[i].toString().trim() : '';

String? _nullIfEmpty(String s) => s.isEmpty ? null : s;

DateTime _parseDate(String yyyymmdd) => DateTime(
      int.parse(yyyymmdd.substring(0, 4)),
      int.parse(yyyymmdd.substring(4, 6)),
      int.parse(yyyymmdd.substring(6, 8)),
    );

int _toInt(String s) =>
    int.tryParse(s) ?? (double.tryParse(s)?.round() ?? 0);

File: supabase/functions/_shared/einvoice_csv_parser.ts
// Pure-TS port of packages/pp_core/lib/src/einvoice_csv_parser.dart.
//
// Intentionally dependency-free (no Deno/npm imports) so it runs identically in
// the Edge Function (Deno) and the Node parity test. Keep in lockstep with the
// Dart original — the parity test (einvoice_csv_parser.test.ts) guards this.

export interface ParsedItem {
  name: string; // 消費明細_品名
  quantity: number; // 消費明細_數量
  unitPrice: number; // 消費明細_單價 (TWD dollars)
  amount: number; // 消費明細_金額 (TWD dollars)
}

export interface ParsedInvoice {
  invoiceNumber: string;
  date: string; // 'YYYY-MM-DD'
  merchantName: string | null; // 賣方名稱
  carrierName: string | null; // 載具自訂名稱
  sellerAddress: string | null; // 賣方地址
  items: ParsedItem[];
}

/// Invoice total in TWD dollars = sum of line amounts.
///
/// The CSV's 發票金額 column is unreliable — for multi-line invoices it mirrors
/// the per-line amount rather than the invoice total — so we always sum the line
/// items (discount lines are negative and net correctly).
export function invoiceTotalDollars(inv: ParsedInvoice): number {
  return inv.items.reduce((sum, i) => sum + i.amount, 0);
}

// 2 letters + 8 digits (digits may be masked with '*' for voided/donated ones).
const INVOICE_NO_RE = /^[A-Z]{2}[0-9*]{8}$/;
const DATE_RE = /^\d{8}$/;

/// Parses a MOF carrier detail CSV into invoices grouped by invoice number.
export function parseEinvoiceCsv(csv: string): ParsedInvoice[] {
  let text = csv;
  if (text.length > 0 && text.charCodeAt(0) === 0xfeff) {
    text = text.slice(1); // strip UTF-8 BOM
  }
  text = text.replace(/\r\n/g, "\n").replace(/\r/g, "\n");

  const rows = parseCsvRows(text);
  if (rows.length === 0) return [];

  const col = columnIndex(rows);

  // Preserve first-seen order while grouping by invoice number.
  const order: string[] = [];
  const groups = new Map<string, ParsedInvoice>();

  for (const row of rows) {
    if (row.length <= col.item) continue;
    const no = cell(row, col.number);
    const dateStr = cell(row, col.date);
    if (!INVOICE_NO_RE.test(no) || !DATE_RE.test(dateStr)) {
      continue; // header, footnotes, or any misaligned row
    }

    let acc = groups.get(no);
    if (!acc) {
      acc = {
        invoiceNumber: no,
        date: parseDate(dateStr),
        merchantName: nullIfEmpty(cell(row, col.merchant)),
        carrierName: nullIfEmpty(cell(row, col.carrier)),
        sellerAddress: nullIfEmpty(cell(row, col.address)),
        items: [],
      };
      groups.set(no, acc);
      order.push(no);
    }

    const name = cell(row, col.item);
    if (name.length === 0) continue;
    acc.items.push({
      name,
      quantity: toNum(cell(row, col.qty), 1),
      unitPrice: toInt(cell(row, col.unitPrice)),
      amount: toInt(cell(row, col.amount)),
    });
  }

  const out: ParsedInvoice[] = [];
  for (const no of order) {
    const inv = groups.get(no)!;
    if (inv.items.length > 0) out.push(inv);
  }
  return out;
}

interface Columns {
  carrier: number;
  date: number;
  number: number;
  merchant: number;
  address: number;
  qty: number;
  unitPrice: number;
  amount: number;
  item: number;
}

function columnIndex(rows: string[][]): Columns {
  const byName = new Map<string, number>();
  for (const row of rows) {
    const trimmed = row.map((c) => c.trim());
    if (trimmed.includes("發票號碼") && trimmed.includes("消費明細_品名")) {
      trimmed.forEach((c, i) => byName.set(c, i));
      break;
    }
  }
  const idx = (name: string, fallback: number) =>
    byName.has(name) ? byName.get(name)! : fallback;
  return {
    carrier: idx("載具自訂名稱", 0),
    date: idx("發票日期", 1),
    number: idx("發票號碼", 2),
    merchant: idx("賣方名稱", 7),
    address: idx("賣方地址", 8),
    qty: idx("消費明細_數量", 10),
    unitPrice: idx("消費明細_單價", 11),
    amount: idx("消費明細_金額", 12),
    item: idx("消費明細_品名", 13),
  };
}

/// Minimal RFC4180-style CSV row splitter (handles quoted fields + escaped
/// quotes). Matches the `csv` package's behaviour for this export's shape.
function parseCsvRows(text: string): string[][] {
  const rows: string[][] = [];
  let field = "";
  let row: string[] = [];
  let inQuotes = false;
  for (let i = 0; i < text.length; i++) {
    const ch = text[i];
    if (inQuotes) {
      if (ch === '"') {
        if (text[i + 1] === '"') {
          field += '"';
          i++;
        } else {
          inQuotes = false;
        }
      } else {
        field += ch;
      }
    } else if (ch === '"') {
      inQuotes = true;
    } else if (ch === ",") {
      row.push(field);
      field = "";
    } else if (ch === "\n") {
      row.push(field);
      rows.push(row);
      row = [];
      field = "";
    } else {
      field += ch;
    }
  }
  if (field.length > 0 || row.length > 0) {
    row.push(field);
    rows.push(row);
  }
  return rows;
}

function cell(row: string[], i: number): string {
  return i < row.length ? row[i].trim() : "";
}

function nullIfEmpty(s: string): string | null {
  return s.length === 0 ? null : s;
}

function parseDate(yyyymmdd: string): string {
  return `${yyyymmdd.slice(0, 4)}-${yyyymmdd.slice(4, 6)}-${yyyymmdd.slice(6, 8)}`;
}

function toInt(s: string): number {
  const t = s.trim();
  if (/^[+-]?\d+$/.test(t)) return parseInt(t, 10);
  const f = parseFloat(t);
  return Number.isNaN(f) ? 0 : Math.round(f);
}

function toNum(s: string, fallback: number): number {
  const t = s.trim();
  if (t === "") return fallback;
  const n = Number(t);
  return Number.isNaN(n) ? fallback : n;
}
